### ISSUE TITLE
dynamically set keycloak

### ISSUE THREAD

**Issue by davirk21** [panz3r/react-keycloak#23]
**Is your feature request related to a problem? Please describe.**
I want to create a login page in my react application, which dynamically set the realm id for the keycloak, but since the keycloak provider wraps my entire application, the application throws error for empty realm string. I don't want to set realm until the user enter realm input on the login page. 

**Describe the solution you'd like**
In the Login page , after entering realm (input) for the keycloak realm onclick a button the keycloak will dynamically initialize with {realm: (whatever input was submitted)}. On a success realm it will redirect to private route, which I should have access to keycloak object with "withKeyclock" HOC.  
**Describe alternatives you've considered**

**Additional context**
************index.js**************
ReactDOM.render(
  KeycloakProvider keycloak={keycloak} keycloakConfig={keycloakConfig}
    Provider store={store}
      Router
        App 
      Router
    Provider
 KeycloakProvider,
  document.getElementById("root")

***************App.js********************
class App extends Component {
  render() {
    return (
      div
        PrivateRoute  
        Switch
          Route exact path="/" component={LoginForm} 
        Switch
      div
    );
  }
}

**Comment by panz3r** [comment]
Hi @davirk21 ,

Thanks for your feedback.

Thinking about your issue, I came up with a solution like this
```js
const keycloakProviderInitConfig = {
  onLoad: 'check-sso',
};

class App extends React.PureComponent {
  constructor(props) {
    super(props);

    const keycloakRealm = localStorage.getItem('kcRealm');

    this.state = {
      keycloak: !!keycloakRealm ? new Keycloak({ realm: keycloakRealm }) : null,
    };
  }

  setupKeycloakWithRealm = keycloakRealm => {
    // ... Validate Keycloak realm...

    // Store Keycloak realm name
    localStorage.setItem('kcRealm', keycloakRealm);

    // Setup a Keycloak instance
    this.setState({
      keycloak: new Keycloak({ realm: keycloakRealm }),
    });
  };

  onKeycloakEvent = (event, error) => {
    console.log('onKeycloakEvent', event, error);
    // On Logout
    if (event === 'onAuthLogout') {
      // Cleanup stored Keycloak realm name
      localStorage.removeItem('kcRealm');

      // Cleanup keycloak instance
      this.setState({
        keycloak: null,
      });
    }
  };

  onKeycloakTokens = tokens => {
    console.log('onKeycloakTokens', tokens);
  };

  render() {
    const { keycloak } = this.state;

    if (!keycloak) {
      return (
        <KeycloakRealmInput onKeycloakRealm={this.setupKeycloakWithRealm} />
      );
    }

    return (
      <KeycloakProvider
        keycloak={keycloak}
        initConfig={keycloakProviderInitConfig}
        onEvent={this.onKeycloakEvent}
        onTokens={this.onKeycloakTokens}
      >
        <AppRouter />
      </KeycloakProvider>
    );
  }
}

export default App;
```

Basically your context is initialised after entering the `Keycloak` realm name in the `KeycloakRealmInput` (which you can implement as an input box or a select), you can then validate it inside `setupKeycloakWithRealm` method and if valid you create a `Keycloak` instance for that realm.

The `localStorage` bits are required to store the `keycloak` realm to allow reinitialising it after a successful login, otherwise the user will be forced to re-enter the realm name after `keycloak` redirect.

Let me know if this helps you.

**Comment by davirk21** [comment]
How do i invoke the keycloak.login method to once the user submit the realm Id from the RealmForm.
Once the Login happen the user will be redirected to the privateRoute 
If the user already authenticated he will be redirected to the privateRoute main
App.js*********
`render() {
    console.log("State", this.state.keycloak);

    const { keycloak } = this.state;
    if (!keycloak) {
      return (
        <RealmForm
          setupKeycloakWithRealm={this.setupKeycloakWithRealm}
          onKeycloakEvent={this.onKeycloakEvent}
        />
      );
    } else {
      return (
        <KeycloakProvider
          keycloak={keycloak}
          initConfig={keycloakProviderInitConfig}
          onEvent={this.onKeycloakEvent}
          onTokens={this.onKeycloakTokens}
        >
          <Router>
            <AppRouter />
          </Router>
        </KeycloakProvider>
      );
    }
  }`

AppRouter.js************

`class AppRouter extends Component {
  componentDidMount() {
    if (this.props.keycloak.authenticated === false) {
      this.props.keycloak.login({
        redirectUri: `${process.env.REACT_APP_BASE_URL}/user/vehicles`,
        kcLocale: "en-US"
      });
    }
  }
  render() {
    console.log("KEYCLOAK IN APP ROUTER", this.props.keycloak);
    if (this.props.keycloak.authenticated) {
      return (
        <div>
          <PrivateRoute path="/user" component={Layout} />
          <PrivateRoute path="/user/demo" component={DemoPage} />
          <PrivateRoute exact path="/user/ingest" component={IngestPage} />
          <PrivateRoute path="/user/dashboard" component={DashboardPage} />
          <PrivateRoute path="/user/profile" component={TenantPage} />
          <PrivateRoute exact path="/user/vehicles" component={VehiclesPage} />
          <PrivateRoute path="/user/routes" component={RoutesPage} />
          <PrivateRoute path="/user/sessions" component={SessionsPage} />
          <PrivateRoute path="/user/edge" component={EdgePage} />
        </div>
      );
    }
  }
}

export default withKeycloak(AppRouter);`

PrivateRoute.js*********
`class PrivateRoute extends Component {
  state = {
    keycloak: null,
    authenticated: true
  };

  render() {
    console.log("KEYKLOACK", this.props.keycloak);
    if (this.props.keycloak.authenticated) {
      localStorage.setItem("realm", this.props.keycloak.realm);
      localStorage.setItem("id_token", this.props.keycloak.idToken);
      localStorage.setItem("access_token", this.props.keycloak.token);
      localStorage.setItem("refresh_token", this.props.keycloak.refreshToken);
      // this.props.history.push("/user/vehicles");
    }

    const { component: Component, ...rest } = this.props;

    return (
      <Route
        {...rest}
        render={props =>
          this.props.keycloak.authenticated ? (
            <Component {...props} />
          ) : (
            <Redirect
              to={{
                pathname: "/",
                state: { from: this.props.location }
              }}
            />
          )
        }
      />
    );
  }
}`

**Comment by panz3r** [comment]
If you need to login the user as soon as `keycloak` is ready you can set `onLoad` init option to `login-required` so you don't need to invoke it explicitly, like this:
```js
const keycloakProviderInitConfig = {
  onLoad: 'login-required',
};
```

**Comment by davirk21** [comment]
Hello Panz3r,

Thank you for the quick responses. Unfortunately I am having a some difficulty setting this up for my application. If you do have some free time please feel free email at davirk21@gmail.com. Maybe if you have free time we can do a video call. I appreciate all the feedback you have given so far.

**Comment by davirk21** [comment]
this.state = {
      keycloak: !!keycloakRealm ? new Keycloak({ realm: keycloakRealm }) : null,
    };

This code require clientID, when I do pass a clientID it goes into an endless loop.

After initialization I try calling the login method through this.state.keycloak.login, but I recieve an error saying .login is undefined. I also tried to call in in my AppRouter component, this works but has a weird bug. So once I call it, I have to comment out the method and do a hard refresh.

In your component how do you maintain keycloak state? Once it does get initialized then should  renders the other condition.

**Comment by panz3r** [comment]
Hi @davirk21 ,

Sorry for the late answer, I had little to no time in the past weeks.

I just uploaded a new branch [sample/dynamic-keycloak](https://github.com/panz3r/react-keycloak/tree/sample/dynamic-keycloak) to test out my initially proposed solution plus some fixes to add missing data on Keycloak setup.
I made some testing and can confirm it to works.

Let me know if this helps you.

**Comment by xtrasmal** [comment]
any idea why keycloak-js 7.0.0 does not work?

**Comment by panz3r** [comment]
Hi @xtrasmal ,
Can you please open a new issue about your problem.
Thanks.

**Comment by jplew** [comment]
@xtrasmal keycloak-js 7.0.0 is buggy. It's a known issue: https://issues.jboss.org/browse/KEYCLOAK-11187

Looks like it was patched though (https://github.com/keycloak/keycloak/pull/6279), we're just waiting on the new release.
